### ISSUE TITLE
`torch.compile` produces different uint8 quantization values for E8M0 `log2 → ceil → clamp → uint8` pipeline compared to eager mode

### ISSUE THREAD

**Issue by himi1008** [pytorch/pytorch#179570]
### 🐛 Describe the bug

`torch.compile` with `inductor` backend produces numerically different uint8-encoded features for a CNN model that uses E8M0 log-domain quantization (`torch.abs() + eps → torch.log2() → torch.ceil() → torch.clamp(-127, 127) → add bias 127 → to(uint8)`). Eager mode is perfectly deterministic (max_var=0 across repeated runs), confirming this is a systematic computation difference introduced by Inductor, not GPU non-determinism.

The `ceil(log2(x))` chain is extremely sensitive to tiny floating-point differences: when a value is near an integer boundary (e.g., `log2(x) ≈ 3.99999` vs `4.00001`), `ceil` rounds in opposite directions, producing uint8 values that differ by 1 per boundary crossing. Inductor may fuse or reorder the log2→ceil chain differently from eager, causing these boundary divergences.

### Minimal reproducer

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class E8M0LogQuantizer(nn.Module):
    """E8M0 quantization via log-domain: abs → log2 → ceil → clamp → uint8."""
    def __init__(self, eps=1e-7, bias=127):
        super().__init__()
        self.eps = eps
        self.bias = bias

    def forward(self, x):
        x_abs = torch.abs(x) + self.eps
        log2_val = torch.log2(x_abs)
        ceil_val = torch.ceil(log2_val)
        clamped = torch.clamp(ceil_val, min=-127, max=127)
        encoded = (clamped + self.bias).to(torch.uint8)
        return encoded


class LogDomainCNN(nn.Module):
    def __init__(self, num_classes=10):
        super().__init__()
        # Feature extraction
        self.conv1 = nn.Conv2d(3, 32, kernel_size=3, padding=1)
        self.bn1 = nn.BatchNorm2d(32)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.bn2 = nn.BatchNorm2d(64)
        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.bn3 = nn.BatchNorm2d(128)
        self.pool = nn.AdaptiveAvgPool2d(4)

        # E8M0 log-domain quantizer
        self.quantizer = E8M0LogQuantizer()

        # Classifier
        self.fc = nn.Linear(128 * 4 * 4, num_classes)

    def forward(self, x):
        x = F.relu(self.bn1(self.conv1(x)))
        x = F.relu(self.bn2(self.conv2(x)))
        x = F.relu(self.bn3(self.conv3(x)))
        x = self.pool(x)

        # E8M0 log-domain quantization
        encoded = self.quantizer(x)

        # Classification from float features
        cls = x.flatten(1)
        logits = self.fc(cls)
        return logits, encoded


# NOTE: CPU — the log2→ceil pipeline bug reproduces on CPU
device = "cpu"
torch.manual_seed(42)
model = LogDomainCNN().to(device).eval()
# Positive-biased input (abs + offset) to ensure valid log2 range
x = torch.abs(torch.randn(4, 3, 64, 64, device=device)) + 0.1

# Eager: deterministic
with torch.no_grad():
    ref_logits, ref_enc = model(x)
    ref_logits2, ref_enc2 = model(x)
print(f"Eager deterministic (logits): {(ref_logits - ref_logits2).abs().max().item():.6e}")
print(f"Eager deterministic (encoded): {(ref_enc.int() - ref_enc2.int()).abs().max().item()}")

# Compiled
torch._dynamo.reset()
compiled = torch.compile(model, backend="inductor")
with torch.no_grad():
    comp_logits, comp_enc = compiled(x)

logit_diff = (ref_logits.float() - comp_logits.float()).abs()
enc_diff = (ref_enc.int() - comp_enc.int()).abs()
print(f"Logits max_diff: {logit_diff.max().item():.6e}")
print(f"Encoded uint8 mismatches: {(enc_diff > 0).sum().item()} / {ref_enc.numel()}")
print(f"Encoded max_diff: {enc_diff.max().item()}")
```

### Behavior summary

| Mode | Result | Notes |
|------|--------|-------|
| Eager | Reference output | Perfectly deterministic across runs (max_var=0) |
| `torch.compile(backend="inductor")` | Different output | uint8 quantized values differ at log2/ceil boundaries |

### Notes

- Eager mode is perfectly deterministic (max_var=0), ruling out non-determinism.
- The bug reproduces on **CPU** — not GPU-specific.
- The `log2 → ceil` pipeline acts as a precision amplifier: even ULP-level float differences near integer boundaries cause `ceil` to round in opposite directions, flipping uint8 values by ±1.
- Multiple boundary crossings across a 128-channel feature map accumulate to larger max_diff values.
- Inductor may optimize the `ceil(log2(x))` chain or fuse it with upstream operations, changing the intermediate precision.

### Error logs

No error — both modes succeed, but produce different uint8 outputs:

```
Eager deterministic (logits): 0.000000e+00
Eager deterministic (encoded): 0
Logits and encoded uint8 values differ systematically under torch.compile
```

### Versions

```
PyTorch version: 2.12.0.dev20260327+cu126
Python: 3.10.12
OS: Ubuntu 22.04.5 LTS (WSL2)
GPU: NVIDIA GeForce RTX 3060 Laptop GPU
CUDA: 12.6
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh @anijain2305

topic: fuzzer

**Comment by morrison-turnansky** [comment]
What order of magnitude are we concerned with? When I run locally I get, which seems well within bounds of tolerable numerical divergence.

Eager deterministic (logits): 0.000000e+00
Eager deterministic (encoded): 0
Logits max_diff: 2.235174e-08
Encoded uint8 mismatches: 0 / 8192
Encoded max_diff: 0

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I attempted to reproduce this issue on PyTorch nightly (`2.13.0.dev20260509+cu130`) on NVIDIA H100. Result: PREDEFINED_MITIGATION.

Issue reports that torch.compile produces different uint8 quantization values vs eager mode in a log2->ceil->clamp->uint8 pipeline. Ran the issue's exact repro 3 times on torch 2.13.0.dev20260509+cu130 (testing on both CPU as in the issue and additionally on CUDA). All 3 runs produced identical, consistent results: 0 uint8 mismatches out of 8192 elements (encoded max_diff = 0) on both CPU and CUDA. Logits max_diff was 2.235174e-08 on CPU and 1.648441e-07 on CUDA - both well below the 1e-6 numerical_tolerance threshold. The reported uint8 boundary divergence does not reproduce on the current nightly. A collaborator (morrison-turnansky) reported the same observation in the issue comments (Logits max_diff: 2.235174e-08, Encoded uint8 mismatches: 0 / 8192). Falls under numerical_tolerance predefined mitigation: float differences are within expected ULP-level tolerance and produce no actual quantization output divergence.

<details>
<summary>Error output</summary>

```
============================================================
Testing on CPU (as in issue)
============================================================
Eager deterministic (logits): 0.000000e+00
Eager deterministic (encoded): 0
Logits max_diff: 2.235174e-08
Encoded uint8 mismatches: 0 / 8192
Encoded max_diff: 0

============================================================
Testing on CUDA
============================================================
Eager deterministic (logits): 0.000000e+00
Eager deterministic (encoded): 0
Logits max_diff: 1.648441e-07
Encoded uint8 mismatches: 0 / 8192
Encoded max_diff: 0

============================================================
Tolerance analysis (CPU - issue's primary device)
============================================================
atol=1e-04: logits PASS (max_diff=2.235174e-08)
atol=1e-05: logits PASS (max_diff=2.235174e-08)
atol=1e-06: logits PASS (max_diff=2.235174e-08)
atol=1e-07: logits PASS (max_diff=2.235174e-08)
atol=1e-08: logits FAIL (max_diff=2.235174e-08)

RESULT: NOT_REPRODUCED - max diffs within numerical tolerance

```
</details>

<details>
<summary>Repro script I used</summary>

```python
import torch
import torch.nn as nn
import torch.nn.functional as F


class E8M0LogQuantizer(nn.Module):
    """E8M0 quantization via log-domain: abs -> log2 -> ceil -> clamp -> uint8."""
    def __init__(self, eps=1e-7, bias=127):
        super().__init__()
        self.eps = eps
        self.bias = bias

    def forward(self, x):
        x_abs = torch.abs(x) + self.eps
        log2_val = torch.log2(x_abs)
        ceil_val = torch.ceil(log2_val)
        clamped = torch.clamp(ceil_val, min=-127, max=127)
        encoded = (clamped + self.bias).to(torch.uint8)
        return encoded


class LogDomainCNN(nn.Module):
    def __init__(self, num_classes=10):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 32, kernel_size=3, padding=1)
        self.bn1 = nn.BatchNorm2d(32)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.bn2 = nn.BatchNorm2d(64)
        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.bn3 = nn.BatchNorm2d(128)
        self.pool = nn.AdaptiveAvgPool2d(4)
        self.quantizer = E8M0LogQuantizer()
        self.fc = nn.Linear(128 * 4 * 4, num_classes)

    def forward(self, x):
        x = F.relu(self.bn1(self.conv1(x)))
        x = F.relu(self.bn2(self.conv2(x)))
        x = F.relu(self.bn3(self.conv3(x)))
        x = self.pool(x)
        encoded = self.quantizer(x)
        cls = x.flatten(1)
        logits = self.fc(cls)
        return logits, encoded


def run_once(device, seed=42):
    torch.manual_seed(seed)
    model = LogDomainCNN().to(device).eval()
    x = torch.abs(torch.randn(4, 3, 64, 64, device=device)) + 0.1

    with torch.no_grad():
        ref_logits, ref_enc = model(x)
        ref_logits2, ref_enc2 = model(x)

    eager_det_logits = (ref_logits - ref_logits2).abs().max().item()
    eager_det_enc = (ref_enc.int() - ref_enc2.int()).abs().max().item()
    print(f"Eager deterministic (logits): {eager_det_logits:.6e}")
    print(f"Eager deterministic (encoded): {eager_det_enc}")

    torch._dynamo.reset()
    compiled = torch.compile(model, backend="inductor")
    with torch.no_grad():
        comp_logits, comp_enc = compiled(x)

    logit_diff = (ref_logits.float() - comp_logits.float()).abs()
    enc_diff = (ref_enc.int() - comp_enc.int()).abs()

    max_logit_diff = logit_diff.max().item()
    enc_mismatches = (enc_diff > 0).sum().item()
    max_enc_diff = enc_diff.max().item()

    print(f"Logits max_diff: {max_logit_diff:.6e}")
    print(f"Encoded uint8 mismatches: {enc_mismatches} / {ref_enc.numel()}")
    print(f"Encoded max_diff: {max_enc_diff}")

    return max_logit_diff, enc_mismatches, max_enc_diff


def main():
    print("=" * 60)
    print("Testing on CPU (as in issue)")
    print("=" * 60)
    cpu_results = run_once("cpu")

    if torch.cuda.is_available():
        print()
        print("=" * 60)
        print("Testing on CUDA")
        print("=" * 60)
        cuda_results = run_once("cuda")

    print()
    print("=" * 60)
    print("Tolerance analysis (CPU - issue's primary device)")
    print("=" * 60)
    max_logit_diff, enc_mismatches, max_enc_diff = cpu_results
    for atol in [1e-4, 1e-5, 1e-6, 1e-7, 1e-8]:
        passes_logits = max_logit_diff < atol
        print(f"atol={atol:.0e}: logits {'PASS' if passes_logits else 'FAIL'} (max_diff={max_logit_diff:.6e})")

    if enc_mismatches == 0 and max_logit_diff < 1e-6:
        print()
        print("RESULT: NOT_REPRODUCED - max diffs within numerical tolerance")
    elif enc_mismatches > 0:
        print()
        print(f"RESULT: REPRODUCED - {enc_mismatches} uint8 mismatches found")
    else:
        print()
        print("RESULT: unclear - inspect output")


if __name__ == "__main__":
    main()

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260509+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260509+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260510+cu130
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.27.0.dev20260510+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>
